Initial draft of JSPI post (#679)
v8/v8.dev: src/blog/jspi.md

Once the flag is enabled, you should be able to run scripts that use JSPI. Below we show how you can use Emscripten to generate a WebAssembly module in C/C++ that uses JSPI. If your application involves a different language, not using Emscripten for example, then we suggest that you look at how the API works you should look at the [proposal](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md).

### Limitations

The chrome implementation of JSPI should already support typical use cases. However it is experimental so there are a few limitations to be aware of:

- Only x64 and arm64 are supported.
- Only JS-to-wasm exports and wasm-to-JS imports are supported. It is not possible for instance to call a re-imported wasm export and suspend when it returns a Promise.
- Each call to a JSPI export runs on a separate stack. For now these stacks have a fixed size. The default stack size in kB can be changed with the V8 flag:

  ```
  –wasm-stack-switching-stack-size
  ```

  From the chrome command line, this would look like:

  ```

>   --js-flags=”--wasm-stack-switching-stack-size=1000”

Reviewer: No curly quotes in this case:

```suggestion
  --js-flags="--wasm-stack-switching-stack-size=1000"
```

Author: good catch!

  ```

- Debugging support is somewhat minimal. In particular, it may be difficult to see the different events happening in the Dev tools panel. Providing a richer support for debugging JSPI applications is on the roadmap.

## A small demo

To see all this working, let’s try a simple example. This C program computes fibonacci in a spectacularly bad way: by asking JavaScript to do the addition, even worse by using JavaScript Promises to do it:[^2]

> ```

Reviewer: ```suggestion
```c
```

Author: done

long promiseFib(long x){
 if(x==0)
   return 0;
 if(x==1)
   return 1;
 return promiseAdd(promiseFib(x-1),promiseFib(x-2));
}
// promise an addition
EM_ASYNC_JS(long, promiseAdd, (long x,long y), {
 return Promise.resolve((k,r)=>{setTimeout(()=> {return k(x+y)},0)});
});
```

The `promiseFib` function itself is a straightforward recursive version of the fibonacci function. The intriguing part (from our point of view) is the definition of `promiseAdd` which does the addition of the two fibonacci halves — using JSPI!.

We use the `EM_ASYNC_JS` Emscripten macro to write down the `promiseFib` function as a Javascript function within the body of our C program. Since addition does not normally involve Promises in JavaScript, we have to force it by using the standard `Promise.resolve` function. In addition, we have to hide the arithmetic behind a `setTimeout` call to make sure that the engine does actually create a Promise that involves the browser’s event loop.

The `EM_ASYNC_JS` macro generates all the necessary glue code so that we can use JSPI to access the Promise’s result as though it were a normal function.

To compile our small demo, we use Emscripten’s `emcc` compiler:[^4]

> ```

Reviewer: ```suggestion
```sh
```

Author: done

We need to use JSPI because APIs like `fetch` are inherently asynchronous in nature, but we want to be able to invoke them from arbitrary places in our application—in particular, from the middle of a call to a function that does not yet exist.

The core idea is to replace a dynamically loaded function with a stub; this stub will first of all load the missing function code, replace itself by the loaded code and then call the newly loaded code with the original arguments. Any subsequent call to the function would go directly to the loaded function. This strategy allows for an essentially transparent approach to dynamically loading code.

The module we are going to load is fairly simple, it contains a function that returns 42:

```
// This is a simple provider of forty-two
#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE long provide42(){
 return 42l;
}
```

which is in a file called `p42.c`, and is compiled using Emscripten without building any ‘extras’:

> ```

Reviewer: ```suggestion
```sh
```

Author: done

   double light = (runTest(runLocal,ix,count)/count)*microSeconds;
   double jsTime = (runTest(runJs,ix,count)/count)*microSeconds;
   double promiseTime = (runTest(runPromise,ix,count)/count)*microSeconds;
   printf("fib(%d) %gμs %gμs %gμs %gμs\n",ix, light, jsTime, promiseTime, (promiseTime-jsTime));
 }
}

EMSCRIPTEN_KEEPALIVE int main(){
 int step =  1;
 int limit = 15;
 int count = 1000;
 runTestSequence(step,limit,count);
 return 0;
}
```

## Appendix B: Listing of `u42.c` and `p42.c`

The `u42.c` C code represents the main part of our dynamic loading example:

```
#include <stdio.h>
#include <emscripten.h>

typedef long (*fooFun)();

// promise a function
EM_ASYNC_JS(fooFun, resolveFun, (), {
 console.log("loading promise42");
 LoadedModule = (await WebAssembly.instantiateStreaming(fetch("p42.wasm"))).instance;
 return addFunction(LoadedModule.exports['provide42']);
}
);

EM_JS_DEPS(funDeps,"$addFunction")

extern fooFun get42;

long stub(){
 get42 = resolveFun();
 return get42();
}

fooFun get42 = stub;

int main(){
 printf("first call p42() = %ld\n",get42());
 printf("second call = %ld\n",get42());
}
```

The p42.c code is the dynamically loaded module.

> ```

Reviewer: To enable syntax-highlighting, specify the programming language following the opening three backticks. E.g. in this case:

```suggestion
```c
```

Consider doing the same for other code snippets throughout the article.

Author: done

#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE long provide42(){
 return 42l;
}
```

<!-- Footnotes themselves at the bottom. -->
## Notes

[^1]:
    For the technically curious, see [the WebAssembly proposal for JSPI](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md) and [the V8 stack switching design portfolio](https://docs.google.com/document/d/16Us-pyte2-9DECJDfGm5tnUpfngJJOc8jbj54HMqE9Y/edit#heading=h.n1atlriavj6v).

[^2]:
    Note: we include the complete program below, in Appendix A.

[^3]: 
    We do not need this flag for our specific example, but you would likely need it for anything bigger.

[^4]:
    Note: you will need a version of Emscripten that is >= 3.1.28.

[^5]:
    The ASYNCIFY=2 option is a reference to the _other_ way of accessing385 asynchronous APIs – using the asyncify feature of Emscripten.

[^6]:

>     the complete program is shown in Appendix B.

Reviewer: ```suggestion
    The complete program is shown in Appendix B.
```

WebAssembly JavaScript Promise Integration (JSPI)” and check the
box. Relaunch as suggested for it to take effect.

You should use at least version `110.0.5473.0` (MacOS) / `110.0.5469.0`
(Windows,Android) / `110.0.5478.4` (Linux) or ChromeOS to get the latest
version of the API. We recommend using the Canary channel to ensure
that any stability updates are applied. In addition, if you wish to
use Emscripten to generate WebAssembly (which we recommend), you
should use a version that is at least `3.1.28`.

It is not yet possible to enable the feature for end users, only to
test it locally by enabling this flag. Eventually we hope to do an
Origin Trial to enable this feature for origins that want to opt in.

Once the flag is enabled, you should be able to run scripts that use
JSPI. Below we show how you can use Emscripten to generate a
WebAssembly module in C/C++ that uses JSPI. If your application
involves a different language, not using Emscripten for example, then
we suggest that you look at how the API works you should look at the
[proposal](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md).

### Limitations

> 

Reviewer 2: Is it worth mentioning here that debugging with devtools is not currently supported?

The chrome implementation of JSPI should already support typical use
cases. However it is experimental so there are a few limitations to be
aware of:

* Only x64 and arm64 are supported.
* Only JS-to-wasm exports and wasm-to-JS imports are supported. It is
  not possible for instance to call a re-imported wasm export and
  suspend when it returns a Promise.
* Each call to a JSPI export runs on a separate stack. For now these
  stacks have a fixed size. The default stack size in kB can be
  changed with the V8 flag:

  ```
  –wasm-stack-switching-stack-size
  ```
  From the chrome command line, this would look like:
  ```
  --js-flags=”--wasm-stack-switching-stack-size=1000”
  ```

## A small demo

To see all this working, let’s try a simple example. This C program
computes fibonacci in a spectacularly bad way: by asking JavaScript to

> do the addition, even worse by using JavaScript `Promise`s to do it:[^2]

Reviewer 2: ```suggestion
do the addition, even worse by using JavaScript Promises to do it:[^2]
```
The formatting seems off with `Promise`s, also inconsistent with the rest of the file that uses Promise/Promises in plain text.

```
long promiseFib(long x){
 if(x==0)
   return 0;
 if(x==1)
   return 1;
 return promiseAdd(promiseFib(x-1),promiseFib(x-2));
}
// promise an addition
EM_ASYNC_JS(long, promiseAdd, (long x,long y), {
 return Promise.resolve((k,r)=>{setTimeout(()=> {return k(x+y)},0)});
});
```

The `promiseFib` function itself is a straightforward recursive
version of the fibonacci function. The intriguing part (from our point
of view) is the definition of `promiseAdd` which does the

> addition of the two fibonacci halves—using JSPI!.

Reviewer 2: ```suggestion
addition of the two fibonacci halves — using JSPI!.
```

## Notes

[^1]:
    For the technically curious, see [the WebAssembly proposal for JSPI](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md) and [the V8 stack switching design portfolio](https://docs.google.com/document/d/16Us-pyte2-9DECJDfGm5tnUpfngJJOc8jbj54HMqE9Y/edit#heading=h.n1atlriavj6v).

[^2]:
    Note: we include the complete program below, in Appendix A.

[^3]: 
    We do not need this flag for our specific example, but you would likely need it for anything bigger.

[^4]:
    Note: you will need a version of Emscripten that is >= 3.1.28.

[^5]:

>     The ASYNCIFY=2 option is a reference to the _other_ way of accessing385 asynchronous APIs – using the asyncify feature of Emscripten.

Reviewer: ```suggestion
    The ASYNCIFY=2 option is a reference to the _other_ way of accessing asynchronous APIs – using the asyncify feature of Emscripten.
```
